verus/lcf/README.md

```diff
@@ -0,0 +1,396 @@
+## Verified Kernel in Dafny
+
+## Purpose 
+
+The purpose of this verified kernel is to encode a RuleSet corresponding to a set of Datalog rules and encode a set of facts( rules without bodies) in order to solve Datalog predicates. The major goal of this kernel is to one day encode a set of rules of criteria for verifiable X.509 web certificate fields and sets of facts corresponding to the fields' values in order to produce a theorem if the proof tree corresponding to these rules and facts is valid. 
+
+## Data types of Kernel
+
+## Const 
+
+Const types are enums that have 3 variants:
+1. Atom(String)
+2. Nat(u64)
+3. Str(String)
+
+Its spec versions for SpecConst are:
+1. Atom(Seq<char>)
+2. Nat(u64)
+3. Str(Seq<char>)
+
+Const is the base type 
+The only function that can be used on the enum is: (in impl DeepView for Const) open spec fn deep_view(&self) -> Self::V
+
+## Subst
+
+Type Subst is a hashmap of Consts:
+type Subst = TmpStringHashMap<Const>;
+
+Its spec verions is SpecSubst which is a Map:
+type SpecSubst = Map<Seq<char>, SpecConst>;
+
+
+## Term
+
+Term types are enums that have 2 variants:
+1. Const(Const)
+2. Var(String)
+
+Its spec versions for SpecTerm are:
+1. Const(SpecConst)
+2. Var(Seq<char>)
+
+Functions that can be used on Term types:
+1. (in impl DeepView for Term) open spec fn deep_view(&self) -> Self::V
+2. (in impl Clone for Term) fn clone(&self) -> (res: Self)
+3. (in impl PartialEq for Term) fn eq(&self, other: &Self) -> (res: bool)
+4. pub open spec fn spec_complete_subst(self, s: SpecSubst) -> bool
+5. pub open spec fn spec_concrete(self) -> bool
+6. pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecTerm)
+7. pub fn term_eq(&self, other: &Self) -> (res: bool)
+8. pub fn complete_subst(self, s: &Subst) -> (res: bool)
+9. pub fn concrete(self) -> (res: bool)
+10. pub fn subst(self, s: &Subst) -> (res: Term)
+
+## Prop
+
+Props are the heads of rules that are used in rulesets for mk_thm.
```

Comment:
Note that `Prop`s don't just appear in rule heads, they are also the bodies of rules---they correspond to the Datalog notion of an "atom" or "predicate".